### PR TITLE
Implement joining over Steam

### PR DESCRIPTION
**Pull request by kaenganxt**
This PR implements joining over the Steam friend list based on #159 

- Check for availability of steam_api64.dll or steam_api.dll and enable Steam feature accordingly
- Set Steam rich presence values when acting as server or client:
  - status: The fact that the user is playing on a server
  - connect: A connection token to enable the "Join Game" button in the friend list
  - steam_player_group: The same token identifying the current server, this will group all players on the same server in the friend list
  - steam_player_group_size: The number of players on the server (Will show "+X other players" in the friend list grouping)

For other friends it looks like this:
grafik
(The text at the bottom says "Playing with 1 other player")
- Implement server tokens
  - Server generates random token
  - Token is sent to joining clients
  - All members provide the token to Steam
  - The token is registered on the API server
  - Joining over Steam then works with the token instead of an IP address and port
- Read command line once in main menu to allow joining over the friend list even if the game is not running
- Listen for Steam rich presence join requests
- Include certain copied files from Steamworks.NET to implement the callback listener
- Make API server port configurable

To test this PR, change the API server port to 4250.

### CODE REVIEW

**Comment by DominicMaas** [comment]
Oh wow!

File: src/csm/Networking/Config/ClientConfig.cs [hunk 1]
```diff
@@ -51,5 +68,8 @@ public ClientConfig()
         ///     Gets the password used to login to the server (if required).
         /// </summary>
         public string Password;
+
+        public bool TokenBased;
+        public string Token;
```
Comment (hibo98) [inline src/csm/Networking/Config/ClientConfig.cs]: Variable documentation

File: src/csm/Commands/Handler/Internal/ClientConnectHandler.cs [hunk 2]
```diff
@@ -23,6 +24,10 @@ protected override void Handle(ClientConnectCommand command)
             }
 
             MultiplayerManager.Instance.PlayerList.Add(command.Username);
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", MultiplayerManager.Instance.PlayerList.Count.ToString());
```
Comment (hibo98) [inline src/csm/Commands/Handler/Internal/ClientConnectHandler.cs]: Make separate methods without magic values

File: src/csm/Commands/Handler/Internal/ClientDisonnectHandler.cs [hunk 3]
```diff
@@ -23,6 +24,11 @@ protected override void Handle(ClientDisconnectCommand command)
 
             TransactionHandler.ClearTransactions(command.ClientId);
             ToolSimulator.RemoveSender(command.ClientId);
+
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", MultiplayerManager.Instance.PlayerList.Count.ToString());
```
Comment (hibo98) [inline src/csm/Commands/Handler/Internal/ClientDisonnectHandler.cs]: Make separate methods without magic values

File: src/csm/Commands/Handler/Internal/ConnectionResultHandler.cs [hunk 4]
```diff
@@ -34,6 +34,12 @@ protected override void Handle(ConnectionResultCommand command)
                 MultiplayerManager.Instance.CurrentClient.ClientPlayer = new Player();
                 MultiplayerManager.Instance.CurrentClient.Status = ClientStatus.Downloading;
                 MultiplayerManager.Instance.CurrentClient.ClientId = command.ClientId;
+                if (CSM.IsSteamPresent)
+                {
+                    SteamHelpers.Instance.SetRichPresence("status", "Playing on a CSM server");
+                    SteamHelpers.Instance.SetRichPresence("connect", "csm=" + command.ServerToken);
+                    SteamHelpers.Instance.SetRichPresence("steam_player_group", command.ServerToken);
```
Comment (hibo98) [inline src/csm/Commands/Handler/Internal/ConnectionResultHandler.cs]: Make separate methods without magic values

File: src/csm/Commands/Handler/Internal/PlayerListHandler.cs [hunk 5]
```diff
@@ -11,6 +12,11 @@ protected override void Handle(PlayerListCommand command)
         {
             MultiplayerManager.Instance.PlayerList.Clear();
             MultiplayerManager.Instance.PlayerList.UnionWith(command.PlayerList);
+
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", MultiplayerManager.Instance.PlayerList.Count.ToString());
```
Comment (hibo98) [inline src/csm/Commands/Handler/Internal/PlayerListHandler.cs]: Make separate methods without magic values

File: src/csm/Helpers/SteamHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+using CSM.Helpers.Steamworks;
+using CSM.Networking;
+using CSM.Networking.Config;
+using CSM.Panels;
+
+namespace CSM.Helpers
+{
+    public class SteamHelpers
+    {
+        public static SteamHelpers Instance;
```
Comment (hibo98) [inline src/csm/Helpers/SteamHelpers.cs]: Make setter private

File: src/csm/Networking/Server.cs [hunk 7]
```diff
@@ -132,6 +143,15 @@ public bool StartServer(ServerConfig serverConfig)
             _hostPlayer.Status = ClientStatus.Connected;
             MultiplayerManager.Instance.PlayerList.Add(_hostPlayer.Username);
 
+            // Set the steam presence 'connect' key. This allows users to click "Join Game" within the steam overlay.
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("status", "Playing on a CSM server");
+                SteamHelpers.Instance.SetRichPresence("connect", "csm=" + ServerToken);
+                SteamHelpers.Instance.SetRichPresence("steam_player_group", ServerToken);
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", "1");
```
Comment (hibo98) [inline src/csm/Networking/Server.cs]: Make separate methods without magic values

File: src/csm/Networking/Server.cs [hunk 8]
```diff
@@ -344,6 +379,10 @@ public void HandlePlayerConnect(Player player)
             Chat.Instance.PrintGameMessage($"Player {player.Username} has connected!");
             MultiplayerManager.Instance.PlayerList.Add(player.Username);
             CommandInternal.Instance.HandleClientConnect(player);
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", MultiplayerManager.Instance.PlayerList.Count.ToString());
```
Comment (hibo98) [inline src/csm/Networking/Server.cs]: Make separate methods without magic values

File: src/csm/Networking/Server.cs [hunk 9]
```diff
@@ -353,6 +392,10 @@ public void HandlePlayerDisconnect(Player player)
             CommandInternal.Instance.HandleClientDisconnect(player);
             TransactionHandler.ClearTransactions(player.NetPeer.Id);
             ToolSimulator.RemoveSender(player.NetPeer.Id);
+            if (CSM.IsSteamPresent)
+            {
+                SteamHelpers.Instance.SetRichPresence("steam_player_group_size", MultiplayerManager.Instance.PlayerList.Count.ToString());
```
Comment (hibo98) [inline src/csm/Networking/Server.cs]: Make separate methods without magic values

File: src/csm/Panels/ConnectionPanel.cs [hunk 10]
```diff
@@ -50,9 +52,18 @@ public override void Start()
...
 
-            // Show Player Pointers
-            _playerPointers = this.CreateCheckBox("Show Player Pointers", new Vector2(10, -270));
+            if (CSM.IsSteamPresent)
+            {
+                // Connected Players window button
+                _inviteFriends = this.CreateButton("Invite Friends", new Vector2(10, -270));
+                _inviteFriends.eventClick += (component, param) =>
+                {
+                    SteamHelpers.Instance.OpenSteamOverlay("friends");
```
Comment (hibo98) [inline src/csm/Panels/ConnectionPanel.cs]: Make separate methods without magic values

**Review by hibo98 — CHANGES_REQUESTED** [review]

**Review by hibo98 — APPROVED** [review]

**Pull request merged**